protobuf_serialization/writer.nim

```diff
@@ -0,0 +1,159 @@
+  stream.write(bytes)
+
+proc writeFixed(stream: OutputStream, fieldNum: int, value: auto) =
+  when sizeof(value) == 8:
+    let wire = Fixed64
+  else:
+    let wire = Fixed32
+  if value.unwrap() == 0:
+    return
+
+  stream.writeProtobufKey(fieldNum, wire)
+  stream.encodeFixed(value)
+
+#stdlib types toProtobuf's. inlined as it needs access to the writeValue function.
+include stdlib_writers
+
+proc writeValueInternal[T](stream: OutputStream, value: T)
+
+proc writeLengthDelimited[T](
+  stream: OutputStream,
+  fieldNum: int,
+  rootType: typedesc[T],
+  fieldName: static string,
+  flatValue: LengthDelimitedTypes
+) =
+  var cursor = stream.delayVarSizeWrite(10)
+  let startPos = stream.pos
+
+  #Byte seqs.
+  when flatValue is CastableLengthDelimitedTypes:
+    if flatValue.len == 0:
+      return
+    stream.write(cast[seq[byte]](flatValue))
+
+  #Standard lib types which use custom converters, instead of encoding the literal Nim representation.
+  elif type(flatValue).isStdlib():
+    stream.stdlibToProtobuf(rootType, fieldName, flatValue)
+
+  #Nested object which even if the sub-value is empty, should be encoded as long as it exists.
+  elif rootType.isPotentiallyNull():
+    writeValueInternal(stream, flatValue)
+
+  #Object which should only be encoded if it has data.
+  elif flatValue is (object or tuple):
+    writeValueInternal(stream, flatValue)
+
+  else:
+    {.fatal: "Tried to write a Length Delimited type which wasn't actually Length Delimited.".}
+
+  if (stream.pos != startPos) or (rootType.isPotentiallyNull()):
+    cursor.finalWrite(newProtobufKey(fieldNum, LengthDelimited) & encodeVarInt(PInt(int32(stream.pos - startPos))))
+  else:
+    cursor.finalWrite([])
+
+
+proc writeFieldInternal[T, R](
+  stream: OutputStream,
+  fieldNum: int,
+  value: T,
+  rootType: typedesc[R],
+  fieldName: static string
+) =
+  static: verifySerializable(flatType(T))
+
+  let flattenedOption = value.flatMap()
+  if flattenedOption.isNone():
+    return
+  let flattened = flattenedOption.get()
+
+  when flattened is VarIntWrapped:
+    stream.writeVarInt(fieldNum, flattened)
+  elif flattened is FixedWrapped:
+    stream.writeFixed(fieldNum, flattened)
+  else:
+    stream.writeLengthDelimited(fieldNum, R, fieldName, flattened)
+
+proc writeField*[T](
+  writer: ProtobufWriter,
+  fieldNum: int,
+  value: T
+) {.inline.} =
+  writer.stream.writeFieldInternal(fieldNum, value, type(value), "")
+
+proc writeValueInternal[T](stream: OutputStream, value: T) =
+  static: verifySerializable(flatType(T))
+
+  let flattenedOption = value.flatMap()
+  if flattenedOption.isNone():
+    return
+  let flattened = flattenedOption.get()
+
+  when flatType(value).isStdlib():
+    stream.writeFieldInternal(1, flattened, type(value), "")
+  elif flattened is (object or tuple):
+    enumInstanceSerializedFields(flattened, fieldName, fieldVal):
+      discard fieldName
+      const fieldNum = getCustomPragmaVal(fieldVal, fieldNumber)
```

Comment:
We should try to provide a helpful error message if the user has failed to add the `fieldNumber` pragma

Reply:
A helpful error message requires getCustomPragmaFixed which I couldn't get working due to symbol resolution errors. I'll open an issue for it.